feat(core): Add filtering, selection and pagination to users

ivov opened the pull request:
https://linear.app/n8n/issue/PAY-646

ivov committed:
feat(core): Add filtering, selection and pagination to users

ivov committed:
Complete implementation

ivov committed:
De-duplicate validation

ivov committed:
Cleanup

ivov committed:
Consolidate feature flags

ivov committed:
More deduplication

ivov committed:
Readability

ivov committed:
Remove unrelated change

ivov committed:
Merge master, resolve conflicts

ivov committed:
Fix MFA issue

ivov committed:
Fix test

ivov committed:
Fix pending tests

krynble commented on packages/cli/src/middlewares/listQuery/dtos/user.filter.dto.ts:
```diff
@@ -0,0 +1,19 @@
+import { Expose } from 'class-transformer';
+import { BaseFilter } from './base.filter.dto';
+
+export class UserFilter extends BaseFilter {
```
We don't have this use case for now but it might be worth filtering by name.

krynble commented on packages/cli/src/middlewares/listQuery/dtos/workflow.filter.dto.ts:
```diff
@@ -21,23 +21,6 @@ export class WorkflowFilter {
-		const result = await validate(this);
-
-		if (result.length > 0) throw new Error('Parsed filter does not fit the schema');
+		return this.toFilter(rawFilter, WorkflowFilter);
```
Loved the clean up 💕

krynble commented on packages/cli/src/controllers/users.controller.ts:
```diff
@@ -354,20 +354,64 @@ export class UsersController {
+			if (skip) findManyOptions.skip = skip;
+
+			if (filter?.isOwner !== undefined) {
+				findManyOptions.relations = ['globalRole'];
```
One consequence of this line is that `select` always brings the `globalRole` regardless of the fields you choose.

ivov replied:
A bit more detail here? The condition only adds `globalRole` if we're filtering by `isOwner`.

`select` with `firstName` only brings that field:

<img width="989" alt="Capture 2023-08-25 at 12 02 23@2x" src="https://github.com/n8n-io/n8n/assets/44588767/747ccb77-6378-4a21-9ad6-781838ef55bb">


`filter` with `isOwner` only brings the owner:

<img width="1051" alt="Capture 2023-08-25 at 12 03 30@2x" src="https://github.com/n8n-io/n8n/assets/44588767/be2475e4-7228-415c-9ad4-507f32dfd13e">

krynble replied:
Ah you're right, I was filtering by `isOwner`. All good.

krynble reviewed (commented):
I've found an odd behavior: if the `select` contains only `["firstName"]` the query fails but if I add `id` to the query it always works.

```
{
    "code": 0,
    "message": "SQLITE_ERROR: no such column: distinctAlias.User_id",
    "stacktrace": "QueryFailedError: SQLITE_ERROR: no such column: distinctAlias.User_id\n    at handler (/Users/omarajoue/Workspace/Projects/n8n/node_modules/.pnpm/typeorm@0.3.12_ioredis@5.2.4_mysql2@2.3.3_pg@8.8.0_sqlite3@5.1.6/src/driver/sqlite/SqliteQueryRunner.ts:113:26)\n    at replacement (/Users/omarajoue/Workspace/Projects/n8n/node_modules/.pnpm/sqlite3@5.1.6/node_modules/sqlite3/lib/trace.js:25:27)\n    at Statement.errBack (/Users/omarajoue/Workspace/Projects/n8n/node_modules/.pnpm/sqlite3@5.1.6/node_modules/sqlite3/lib/sqlite3.js:15:21)"
}
```
<img width="1239" alt="Screenshot 2023-08-25 at 11 30 37" src="https://github.com/n8n-io/n8n/assets/219272/3bfd2b95-bffd-4fe0-84e4-6cb50ff7ca1c">

krynble commented:
Also one odd behavior: I have a user whose email is "omar+1@n8n.io" and this doesn't work when filtering

ivov committed:
Support `firstName` and `lastName` filters

ivov commented:
> Also one odd behavior: I have a user whose email is "[omar+1@n8n.io](mailto:omar+1@n8n.io)" and this doesn't work when filtering

You're not using other options? Maybe a screenshot? Unable to reproduce for now:

<img width="988" alt="Capture 2023-08-25 at 11 56 39@2x" src="https://github.com/n8n-io/n8n/assets/44588767/339c4592-e0cf-4a8a-bc2a-94c1c74abf6b">

ivov commented:
> I've found an odd behavior: if the `select` contains only `["firstName"]` the query fails but if I add `id` to the query it always works.
> 
> ```
> {
>     "code": 0,
>     "message": "SQLITE_ERROR: no such column: distinctAlias.User_id",
>     "stacktrace": "QueryFailedError: SQLITE_ERROR: no such column: distinctAlias.User_id\n    at handler (/Users/omarajoue/Workspace/Projects/n8n/node_modules/.pnpm/typeorm@0.3.12_ioredis@5.2.4_mysql2@2.3.3_pg@8.8.0_sqlite3@5.1.6/src/driver/sqlite/SqliteQueryRunner.ts:113:26)\n    at replacement (/Users/omarajoue/Workspace/Projects/n8n/node_modules/.pnpm/sqlite3@5.1.6/node_modules/sqlite3/lib/trace.js:25:27)\n    at Statement.errBack (/Users/omarajoue/Workspace/Projects/n8n/node_modules/.pnpm/sqlite3@5.1.6/node_modules/sqlite3/lib/sqlite3.js:15:21)"
> }
> ```
> 
> <img alt="Screenshot 2023-08-25 at 11 30 37" width="1239" src="https://user-images.githubusercontent.com/219272/263230085-3bfd2b95-bffd-4fe0-84e4-6cb50ff7ca1c.png">

The screenshot shows taking 2 but there can only be 1 owner. 

Should `take=n` fail if there is not enough `n`? Or return as many as there are?

krynble commented:
> > Also one odd behavior: I have a user whose email is "[omar+1@n8n.io](mailto:omar+1@n8n.io)" and this doesn't work when filtering
> 
> You're not using other options? Maybe a screenshot? Unable to reproduce for now:
> 
> <img alt="Capture 2023-08-25 at 11 56 39@2x" width="988" src="https://user-images.githubusercontent.com/44588767/263236982-339c4592-e0cf-4a8a-bc2a-94c1c74abf6b.png">

Hum odd, does not seem to work for me, see the screenshots below

![Screenshot 2023-08-25 at 12 12 14](https://github.com/n8n-io/n8n/assets/219272/4beecd4c-cd04-4f56-a69f-4266187c5ff8)
Above you can see that there's a user with `omar+2@n8n.io` and no filters in the query

![Screenshot 2023-08-25 at 12 13 32](https://github.com/n8n-io/n8n/assets/219272/1efd80f0-f76b-47ba-bb11-a91d53d49573)
Filtering for `omar@n8n.io` works

![Screenshot 2023-08-25 at 12 13 51](https://github.com/n8n-io/n8n/assets/219272/b092dc8e-4f4b-4ae8-ab95-a6bd4e2f2871)
Filtering by `omar+2@n8n.io` is not working

ivov commented:
Re: `omar+2@n8n.io` not being found with `filter`

Very odd, cannot reproduce yet. Tested with a DB from scratch and recreated users and all working:
- https://share.cleanshot.com/d7tT3c2d
- https://share.cleanshot.com/510P8lwG
- https://share.cleanshot.com/GkG4gT59

krynble commented:
> > I've found an odd behavior: if the `select` contains only `["firstName"]` the query fails but if I add `id` to the query it always works.
> > ```
> > {
> >     "code": 0,
> >     "message": "SQLITE_ERROR: no such column: distinctAlias.User_id",
> >     "stacktrace": "QueryFailedError: SQLITE_ERROR: no such column: distinctAlias.User_id\n    at handler (/Users/omarajoue/Workspace/Projects/n8n/node_modules/.pnpm/typeorm@0.3.12_ioredis@5.2.4_mysql2@2.3.3_pg@8.8.0_sqlite3@5.1.6/src/driver/sqlite/SqliteQueryRunner.ts:113:26)\n    at replacement (/Users/omarajoue/Workspace/Projects/n8n/node_modules/.pnpm/sqlite3@5.1.6/node_modules/sqlite3/lib/trace.js:25:27)\n    at Statement.errBack (/Users/omarajoue/Workspace/Projects/n8n/node_modules/.pnpm/sqlite3@5.1.6/node_modules/sqlite3/lib/sqlite3.js:15:21)"
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > <img alt="Screenshot 2023-08-25 at 11 30 37" width="1239" src="https://user-images.githubusercontent.com/219272/263230085-3bfd2b95-bffd-4fe0-84e4-6cb50ff7ca1c.png">
> 
> This is caused by taking more entities than there are available.
> 
> Should `take=n` fail if there are not enough `n`? Or return as many as there are?

I'd say yes, that's what usually pagination systems should do IMO. If your search doesn't contain more results, it just won't return any

ivov committed:
Fix combination of options requiring auxiliary fields

ivov commented:
> I've found an odd behavior: if the select contains only ["firstName"] the query fails but if I add id to the query it always works.

Fixed in [`8ae8522`](https://github.com/n8n-io/n8n/pull/6994/commits/8ae852239121803e9c921d22ba39820338567d17), added test.

ivov committed:
Merge master, resolve conflicts

ivov commented:
Re: omar+2@n8n.io not being found with filter

As discussed the root cause was that the server expects URI-encoded JSON, so the `+` was being parsed as whitespace, so no result was being found. Postman and the browser do not auto-encode, but Insomnia does.

krynble commented:
Thanks for the fixes. Found another weird issue here.

![Screenshot 2023-08-25 at 14 25 32](https://github.com/n8n-io/n8n/assets/219272/437f3153-fdfd-4129-bf99-65e7306eab85)

Using only `take` now results in some weird result

ivov committed:
Fix `take` without selection

ivov committed:
Remove early return

ivov committed:
Fix lint

ivov committed:
Cleanup

ivov commented:
> Found another weird issue here.

Thanks for rechecking. Fixed and added more tests.

ivov committed:
Add clarifying comment

krynble commented on packages/cli/test/integration/users.controller.test.ts:
```diff
@@ -0,0 +1,248 @@
+};
+
+describe('GET /users', () => {
+	test('should return all users', async () => {
```
One thing that might be worth adding is a test for retrieving users with non-owners so we can make sure it always returns, since the `GET` endpoint should be public.

krynble reviewed (commented):
One thing that seems to be a bit odd is that using only `skip` without take does not seem to work, not sure if intended.

Other than this, I think it's worth adding that last test before merging.

ivov commented:
> One thing that seems to be a bit odd is that using only `skip` without take does not seem to work, not sure if intended.

That's intended, `skip` requires `take`. I'll add an error if it's clearer.

> One thing that might be worth adding is a test for retrieving users with non-owners so we can make sure it always returns, since the GET endpoint should be public.

That case is covered here, or do you mean something else?

![Capture 2023-08-28 at 15 05 24@2x](https://github.com/n8n-io/n8n/assets/44588767/d4c803ae-5969-42c9-bc74-430bf50a4d82)

ivov committed:
Switchn to throw on take-less skip

krynble reviewed (approved):
LGTM!

janober commented:
Got released with [`n8n@1.5.1`](https://github.com/n8n-io/n8n/releases/tag/n8n%401.5.1)
